toolchain/lex/string_literal.cpp

```diff
@@ -341,9 +351,9 @@ static auto ExpandAndConsumeEscapeSequence(LexerDiagnosticEmitter& emitter,
 // Expand any escape sequences in the given string literal.
 static auto ExpandEscapeSequencesAndRemoveIndent(
     LexerDiagnosticEmitter& emitter, llvm::StringRef contents, int hash_level,
-    llvm::StringRef indent) -> std::string {
-  std::string result;
-  result.reserve(contents.size());
+    llvm::StringRef indent) -> StringLiteral::ComputedValue {
+  auto result = std::make_unique<std::string>();
```

Comment:
`unique_ptr<string>` is a very inefficient representation for immutable string data; for large strings, this requires two heap representations rather than one, and for small strings that fit comfortably in the small string buffer, it will overallocate.

If we want to guarantee to put the data into a heap allocation (so we can form stable pointers into it), perhaps we could use a `llvm::SmallString` here, and copy into a `std::unique_ptr<char[]>` when producing the result.

Reply:
How much of an issue is space efficiency for strings? My rough understanding for Clang had been that most of the memory consumption was due to the AST, rather than string literals. I'm hesitant to favor additional string copies over the space efficiency, but are string literals actually a big space consumer? Put differently, while I think the heap allocation occurs either way, moving it from `std::string` (or `llvm::SmallString` if you prefer) to `char[]` removes the overhead of the string type but adds the cost of a copy.

I wouldn't expect cache density to be affected unless we get into a more esoteric storage, such as storing strings sequentially in a single vector (which, could get into more copy overhead, though could avoid most of it by preallocating a buffer the size of the source file). Which we could do: and it actually wouldn't be that bad because it could take the buffer as input, and the StringRef could just point at subsections of a bigger SmallString. Which, hmmmmm. Is that tempting to you?

Possibly relevant note, chandlerc had said he's not focused on optimizing string lexing because they're too rare in code (about 1% of lines IIRC), and thus improvements wouldn't be significant to his lexer performance goals. That said, that works both ways: slightly costlier lexing is more okay because they're rare, if memory efficiency is important.

Reply:
My suggestion would reduce the number of heap allocations in most cases (because in the case where the string fits in the `SmallString` the only allocation would be the final one), and remove a level of indirection, in addition to reducing the wasted space. Maybe it's not a huge deal, but it seems needless and wasteful to store a pointer to a pointer to an overallocated buffer here.

Reply:
I'd like to discuss this with you a little. To be sure I'm understanding correctly:

- SmallString will not heap allocate for strings under the chosen size, likely 64 bytes.
- The current logic only runs for "complex" strings that are either multi-line or have escapes.
- Thus, one heap allocation would be avoided for short strings which are also complex.

In turn, string copies would always occur, but it sounds like you're saying it's not worth considering as part of trade-offs. Maybe we can discuss this further?

Reply:
Per discussion, here's an implementation using BumpPtrAllocator and SmallString.

I also worked on the benchmarks, trying to break them down a little further to see the performance impact of the different approach.

BumpPtr and SmallString (memcpy):

```
--------------------------------------------------------------------------------------------
Benchmark                                                  Time             CPU   Iterations
--------------------------------------------------------------------------------------------
BM_ValidString_Simple                                  31218 ns        31217 ns        22792
BM_ValidString_Multiline                               31077 ns        31076 ns        21898
BM_ValidString_MultilineDoubleQuote                    30713 ns        30708 ns        22088
BM_ValidString_Raw                                     30667 ns        30667 ns        22109
BM_IncompleteWithEscapes_Simple                        45669 ns        45668 ns        15482
BM_IncompleteWithEscapes_Multiline                     37472 ns        37469 ns        17954
BM_IncompleteWithEscapes_MultilineDoubleQuote          37836 ns        37835 ns        18471
BM_IncompleteWithEscapes_Raw                           92009 ns        91993 ns         7389
BM_ComputeValue_NoGenerate_Short                        20.2 ns         20.2 ns     34602757
BM_ComputeValue_NoGenerate_Long                         3055 ns         3054 ns       226271
BM_ComputeValue_WillGenerate_Short                      53.5 ns         53.5 ns     13017676
BM_ComputeValue_WillGenerate_Long                      15560 ns        15559 ns        44032
BM_ComputeValue_WillGenerate_Multiline                 15226 ns        15226 ns        43788
BM_ComputeValue_WillGenerate_MultilineDoubleQuote      15406 ns        15406 ns        45770
BM_ComputeValue_WillGenerate_Raw                       15400 ns        15400 ns        46383
```

Versus unique_ptr and SmallString (no memcpy):

```
--------------------------------------------------------------------------------------------
Benchmark                                                  Time             CPU   Iterations
--------------------------------------------------------------------------------------------
BM_ValidString_Simple                                  30177 ns        30176 ns        23630
BM_ValidString_Multiline                               30897 ns        30897 ns        22199
BM_ValidString_MultilineDoubleQuote                    30905 ns        30903 ns        22953
BM_ValidString_Raw                                     30965 ns        30962 ns        22308
BM_IncompleteWithEscapes_Simple                        45130 ns        45122 ns        15771
BM_IncompleteWithEscapes_Multiline                     40566 ns        40563 ns        17345
BM_IncompleteWithEscapes_MultilineDoubleQuote          41453 ns        41445 ns        17584
BM_IncompleteWithEscapes_Raw                           96260 ns        96250 ns         6956
BM_ComputeValue_NoGenerate_Short                        21.1 ns         21.1 ns     34254556
BM_ComputeValue_NoGenerate_Long                         3149 ns         3149 ns       225301
BM_ComputeValue_WillGenerate_Short                      50.0 ns         50.0 ns     14118668
BM_ComputeValue_WillGenerate_Long                       9304 ns         9304 ns        75569
BM_ComputeValue_WillGenerate_Multiline                  9272 ns         9270 ns        75587
BM_ComputeValue_WillGenerate_MultilineDoubleQuote       9333 ns         9333 ns        76496
BM_ComputeValue_WillGenerate_Raw                        9340 ns         9339 ns        74959
```

I understand that the benchmark may not be truly representative of runtime (the string size tested of 10KB should not be considered typical). For a short string it's similar performance.